### ISSUE TITLE
feature propsal - scan multiple branches

### ISSUE THREAD

**Issue by thatsk** [gocd/gocd#6123]
##### Issue Type
- Feature proposal

##### Summary
gocd config repositories should support regex for scanning multiple branches.

![image](https://user-images.githubusercontent.com/15324013/55874301-92b12c80-5baf-11e9-8a93-dd40dc27853a.png)

##### Environment
gocd  19.1.0

**Comment by arvindsv** [comment]
I've been playing around with an idea to support branches with pipelines as code, in my head. I'm putting it out there as a nascent idea to be shot down / improved.


Assumption 1: Some kind of plugin-level templating mechanism (as in https://github.com/gocd/gocd/issues/5675) exists in the config-repo plugins.

#### Setup:
1. Imagine a pipeline defined in YAML / JSON with a single material (for now).
2. Assume that the material is marked, in some way, to need support for branches.

#### Example:

```yaml
track_branches_for: https://www.github.com/gocd/gocd
  patterns: pull/.*/head

  template:
    {{ foreach branch in CONTEXT['branches'] }}
      pipeline-${branch}:
        tag: mytag1
        group: mygroup
        materials:
          mygit:
            git: https://www.github.com/gocd/gocd
            branch: ${branch}
        stages:
          - stage1:
            ...
          
      pipeline-downstream-${branch}:
        tag: mytag1
        group: mygroup
        materials:
          mygit:
            git: https://www.github.com/gocd/gocd
            branch: ${branch}
          myupstream:
            pipeline: pipeline-${branch}
            stage: stage1
        stages:
          mystage:
            ...
    {{ end }}
```

#### Work:
  - The plugin would need to be aware of this and be able to understand branches + filter them.
  - The plugin would need to provide that as context to the template (see assumption that there is a templating mechanism).
  - Depending on how it is implemented, once a PR is merged or a branch is deleted, the context won't include that branch and the pipeline automatically goes away from GoCD.

#### Benefits of this approach:

1. Flexibility: You can use this to build whole VSMs for branches, if you want.
2. Proper branch support: Unlike with branches and PRs using the GoCD PR plugin, there won't be only one pipeline across all branches.
3. With some appropriate use of a concept such as tags, etc. all pipelines tagged with the same "tag" can be shown together in a dashboard, for visualization reasons, to reduce the clutter of having some many pipelines per branch.
4. GoCD core does not need to change (introducing branch support in GoCD core introduces a can of worms - such as, what happens if there are multiple materials tracking branches, how do you templatize this properly, etc).

#### More considerations:
1. Cruft left behind in the DB, after pipeline goes away. Not much we can do with this.
2. Pipeline goes away once a branch is deleted / PR is merged. Maybe not ideal, but probably not a big deal in many cases.

**Comment by arvindsv** [comment]
Just for the record, I was expecting tag to work like this (somewhat):

![bitmap](https://user-images.githubusercontent.com/172235/58572660-18de0c80-820a-11e9-9d3a-7f1e2474fa77.png)

**Comment by arvindsv** [comment]
If we're considering this, I'd just explore it as a part of the Groovy plugin, though. It already comes with "templating" since it's a full-fledged language. /cc @ketan

**Comment by moritz** [comment]
This solution sounds like a very good 80/20 approach, with some obvious limitations. For example, the character set for branch names isn't restricted the same way as GoCD's pipeline names.

Also as discussed in #5675, I'd recommend using  data template language, not a text-based template, to avoid the failure mode where the output of a template isn't valid JSON/YAML (just think of trailing commas being forbidden in JSON, a nightmare to produce correctly from a template).

**Comment by arvindsv** [comment]
> This solution sounds like a very good 80/20 approach, with some obvious limitations. For example, the character set for branch names isn't restricted the same way as GoCD's pipeline names.

True. It doesn't work even in the example I posted. `pull/.*/head` won't be valid. There would need to be some kind of extraction. But, it was more about the idea.

> Also as discussed in #5675, I'd recommend using data template language, not a text-based template, to avoid the failure mode where the output of a template isn't valid JSON/YAML (just think of trailing commas being forbidden in JSON, a nightmare to produce correctly from a template).

I agree. I've seen it be a big problem in [GoCD's helm chart templates](https://github.com/helm/charts/blob/c0af598986f8fe47d15260140e4e1e6578a53279/stable/gocd/templates/gocd-agent-deployment.yaml#L56) as well.

**Comment by marques-work** [comment]
I've spiked this out with the Groovy DSL plugin. See:

- [GoCD changes](https://github.com/marques-work/gocd/tree/branch-support-spike)
- [Groovy DSL changes](https://github.com/marques-work/gocd-groovy-dsl-config-plugin/tree/spikey-spikey-prod-unlikely)

### Caveats for this spike:

  - only Git; keeps it simple
  - Groovy DSL plugin only; for simplicity/reasons outline by @arvindsv
  - to hell with performance
  - a lot of ~cheating~ magic tricks
  - did not investigate UI grouping/display of auto-generated pipelines; they look just like any other pipeline, and in no particular order

### Implementation

1. `ConfigMaterialUpdateListener` always sends `parse-directory`, even if there are no changes
2. Bump the periodicity to 1 min; keep your trousers on and see the caveat about performance
3. New domain concepts/structures for CR plugins: `BranchedWorkflows`, `BranchMatcher`, `BranchContext`
4. During `parse-directory` we iterate through `BranchMatcher` objects and collect/filter remote branch names
    - we cheat here because the Groovy plugin will just shell out to `git ls-remote --refs`
    - really, this should be a message from plugin => GoCD server to collect branches so as to abstract and provide this capability to any plugin in a consistent way; GoCD server should handle the implementation details
    - yes, this might make the parse-directory a little chatty; should think about implications and _All the Things_ that can go wrong
5. Iterate over the collection of matched branch names with a closure for each match to allow the user to define pipeline(s) using the branch info from bound closure parameters
    - provide raw branch name, sanitized branch name (pipeline name compatible), and a prebuilt material object for that branch
6. Return the denormalized partial config JSON to GoCD for validation and merge

### Questions:

1. Github PRs live forever in the repo (i.e., as `refs/pull/*`), so need a way to only consider open/active PRs
2. How to handle `parse-content`? Branch support makes this network-chatty, potentially
3. Will this have any implications for pipeline export? (probably not -- it just converts the merged config into syntax)


### Example Syntax

```groovy
package branch_support_spike

import cd.go.contrib.plugins.configrepo.groovy.dsl.GoCD

GoCD.script {
  branches {
    matching {
      url = "https://url.domain/path/to/repo.git"
      pattern = ~/^refs\/[^\/]+\/.+/

      onMatch { ctx ->
        pipeline("pr-${ctx.branchSanitized}") {
          group = "example-group"

          // preconfigured material available in binding context
          materials { add(ctx.repo) }

          stages {
            stage('tests') {
              jobs {
                job('units') {
                  tasks { bash { commandString = 'whoami' } }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by arvindsv** [comment]
> Github PRs live forever in the repo (i.e., as refs/pull/*), so need a way to only consider open/active PRs

@marques-work I think this will be a challenge. I don't think GitHub adds enough metadata into the git tree to allow us to know this. I mean, if it's a standard merge, we could do things like this:

```
git rev-list --first-parent SHA..origin/master
```

... to figure out whether a branch/PR with a given SHA is merged into master. But, this will probably not work for "Squash merge" which they allow now. If that is the case, we might need to talk to the GitHub API itself and parse that information, which brings along all the problems such as API rate limiting, etc.

This prototype of yours will work really well for a normal git model, where branches are opened, merged and possibly deleted.

**Comment by marques-work** [comment]
@arvindsv yeah, it s difficult. I really only think the reliable way is to consult github's API, but that is specific to GH. A workaround that may be good enough for 75% of cases is a sorting option + limit (e.g., descending numeric sort, last 10).

**Comment by marques-work** [comment]
### Productionalizing Challenges (at least for now)

1. We can't have GoCD cloning each time with `parse-directory` regardless of change; it otherwise may not scale for large instances. We need _some_ way of reusing what has already been parsed to generate new pipeline configs when new branches are created but when the config repo's material itself has not changed.
2. Provide a branch scanning extension endpoint for configrepo plugins

**Comment by marques-work** [comment]
### More Thoughts

I've implemented a branch polling/filtering message to GoCD so that any configrepo plugin can ask for branches matching a regex string.

This won't solve any scaling problems (as it can make a network call to the SCM provider each time a configrepo plugin asks to reevaluate its template tied to the branch pattern.

It also doesn't solve the problem of only retrieving open GitHub PRs.

Will continue/update this comment later on possible solutions tomorrow. It's late here in CA.

**Comment by marques-work** [comment]
PR tracking via Webhooks:

- **So far:** Implemented Webhook endpoint to consume GH PR events
    - controller endpoint results in invoking `MaterialService#{ignoreRef(),allowRef()}` to add or remove pull requests from a cached list of open PRs (plaintext file whitelist of PR refs, 1 per line, located in `.git`); alternately could be stored in DB.
- **Open problem:** Need a way to sync the full list of open PRs because:
    - Need to seed initial list of PRs so we know which refs are closed PRs (and can ignore during branch enumeration) before any webhooks are delivered
    - Users will probably want a "hard-refresh" because webhook deliveries can fail, resulting in out-of-sync PR lists
    - Users may clean the filesystem of cached repo checkouts
- **Possible solutions:**
    - Service-specific webhook doesn't add/remove single refs in the cached file; instead, always refreshes the whole list from the API
    - Create special PR-tracking material types for GitHub/Gitlab/BitBucket that fetch PRs and branches using APIs instead of `git ls-remote --refs`; webhooks would be an optional optimization (alternative to polling).

@arvindsv @maheshp ☝️

**Comment by arvindsv** [comment]
Cool. Thank you for mentioning the current challenges. I'm trying to think of what a "good enough" start is, which can be polished into a more complete solution later. Maybe that is the idea of refreshing the whole list? However, that can involve a lot of paging and requests. 

Maybe we can use https://developer.github.com/v3/activity/events/#list-repository-events to update missing events?

**Comment by arvindsv** [comment]
The PR tracking material seems like it would be a bit much. It might be cleaner, but I'd like to look into it only if everything else doesn't make sense.

**Comment by marques-work** [comment]
Question for myself to think about (but any voluntary thoughts also welcome from others):

Should `parse-content` behave similarly? One of the original intents of `parse-content` was to give the ability for devs to do frequent checks for syntactical and structural correctness as well as evaluating mergeability with the main GoCD config before actually committing. Thus, there's a possibility that `parse-content` can be called more frequently while making changes or authoring a new config repo definition file. As this feature could make real calls to APIs or SCM hosts, is this something to worry about (e.g., network latency and hitting API limits)?

Alternatives:

- Make branch matching a no-op (i.e., ignore it)
- Stub out random values for ref retrieval and avoid real network calls (benefit: allows `parse-content` to vet the template structure while sidestepping downsides)

**Comment by arvindsv** [comment]
I'd prefer random-values, called out as random values, so that error messages show "branch-random-name-for-checking-1" and not "0609bbae3aec9".

**Comment by marques-work** [comment]
Hey all, thought I'd give a little update so it doesn't look like nothing is happening on this front. I'm actively working on this feature, but there's a number of moving parts to coordinate, so it's taken me quite a bit of time.

Non-exhaustive list of pieces (I'm sure there's a few more that I've not thought of, but these are the major bits):

- [x] building out git provider support (e.g., github, bitbucket, etc) in the groovy plugin
- [x] providing a way to securely bind sensitive information (e.g., API auth keys) into groovy templates (user-defined configuration property resolution within a groovy pipeline definition file)
- [x] change GoCD server config-repo parsing and merging behavior to absorb new parsed results that result from changes outside of the config-repo material branch (e.g., newly opened/merged PRs)
- [ ] allow `git` materials to fetch/checkout arbitrary refs (**WIP**)
- [ ] webhook triggers on PR creation/merge to reparse config-repos
- [ ] cache remote refnames to avoid excessive API requests to providers from config-repo material polling (to be used in conjunction with webhook triggers)
- [ ] UI widgets to configure some of these new options (partially done)
- [ ] fixing yaml and json plugins to work with new extension changes; this does not entail supporting branch/PR scanning but only to make them not blow up

I can't guarantee how soon this work will be completed for public consumption, but I do anticipate a basic working version that the core team can dogfood internally within a couple of weeks.

**Comment by marques-work** [comment]
> I'd prefer random-values, called out as random values, so that error messages show "branch-random-name-for-checking-1" and not "0609bbae3aec9".

@arvindsv Yep, that's taken care of :) -- https://github.com/gocd-contrib/gocd-groovy-dsl-config-plugin/blob/ffae7c6e58bf5bf39441c15d9037c54df5321bb9/resolvers/src/main/java/cd/go/contrib/plugins/configrepo/groovy/resolvers/Branches.java#L69-L83

**Comment by marques-work** [comment]
## Update September 2020 (again)

Good news: [We've released the first version of the PR support in the Groovy DSL plugin](https://github.com/gocd-contrib/gocd-groovy-dsl-config-plugin/releases/tag/v1.1.0-233).

You'll need to be running on GoCD 20.8.0 in order to use this plugin (relies on some new APIs). It provides PR scanning for providers that support pull/merge requests (namely GitHub, GitLab, Bitbucket, and self-hosted Bitbucket), and basic branch/tag scanning for vanilla `git` repositories.

This plugin release also includes built-in pipeline/stage status notifications (configurable via DSL) so you can send build statuses back to your `git` provider to associate builds with pull/merge requests.

The to-do list in the previous comment still stands, and will be released some time in the future:

- [ ]  Webhook support through `autoUpdate = false` configuration
- [ ]  Add configurable description to pipeline group and include a DSL for this (...maybe? this will take some thought as to how to model it properly in the DSL)

**Comment by marques-work** [comment]
This is pretty much implemented fully now using the groovy dsl plugin.

The one nice-to-have that I didn't do is `Add configurable description to pipeline group and include a DSL for this` but I don't think this one feature should hold this issue open any further. If this feature is really wanted/needed, we can create a new issue for it. But alas, this PR/branch support is available.

Closing this finally.

**Comment by thatsk** [comment]
@arvindsv Is there any implementation that will going to happen on yaml config plugin same like groovy DSL Plugin

**Comment by arvindsv** [comment]
@thatsk I don't think so. That was the original idea, but as far as I know, when @marques-work looked into doing that, there were too many issues with going down that route.
